cpp/src/preprocessing/spectral/spectral_embedding.cu

```diff
@@ -0,0 +1,241 @@
+
+  cuvs::neighbors::brute_force::search_params search_params;
+
+  cuvs::neighbors::brute_force::search(
+    handle, search_params, index, nums, d_indices.view(), d_distances.view());
+
+  // Create a COO matrix for the KNN graph
+  raft::sparse::COO<float> knn_coo(stream, n_samples, n_samples);
+
+  // Resize COO to actual nnz
+  size_t nnz = n_samples * k_search;
+  knn_coo.allocate(nnz, n_samples, false, stream);
+
+  auto knn_rows = raft::make_device_vector<int>(handle, nnz);
+  auto knn_cols = raft::make_device_vector<int>(handle, nnz);
+
+  thrust::transform(thrust::device,
+                    d_indices.data_handle(),
+                    d_indices.data_handle() + nnz,
+                    knn_cols.data_handle(),
+                    [] __device__(int64_t x) -> int { return static_cast<int>(x); });
+
+  thrust::transform(thrust::device,
+                    thrust::counting_iterator<int>(0),
+                    thrust::counting_iterator<int>(nnz),
+                    knn_rows.data_handle(),
+                    [=] __device__(int64_t i) { return static_cast<int>(i / k_search); });
+
+  // Copy COO data to device
+  raft::copy(knn_coo.rows(), knn_rows.data_handle(), nnz, stream);
+  raft::copy(knn_coo.cols(), knn_cols.data_handle(), nnz, stream);
+  raft::copy(knn_coo.vals(), d_distances.data_handle(), nnz, stream);
+
+  raft::sparse::COO<float> coo_no_zeros(stream);  // Don't pre-allocate dimensions
+  raft::sparse::op::coo_remove_zeros<float>(&knn_coo, &coo_no_zeros, stream);
+
+  // binarize to 1s
+  thrust::fill(thrust::device, coo_no_zeros.vals(), coo_no_zeros.vals() + coo_no_zeros.nnz, 1.0f);
+
+  // Create output COO for symmetrized result - create unallocated COO
+  raft::sparse::COO<float> sym_coo1(stream);  // Don't pre-allocate dimensions
+
+  // // Define the reduction function with the correct signature
+  auto reduction_op = [] __device__(int row, int col, float a, float b) {
+    // Only use the values, ignore row/col indices
+    return 0.5f * (a + b);
+  };
+
+  // Symmetrize the matrix
+  raft::sparse::linalg::coo_symmetrize(&coo_no_zeros, &sym_coo1, reduction_op, stream);
+
+  raft::sparse::op::coo_sort<float>(&sym_coo1, stream);
+
+  raft::sparse::COO<float> sym_coo(stream);  // Don't pre-allocate dimensions
+  raft::sparse::op::coo_remove_zeros<float>(&sym_coo1, &sym_coo, stream);
+
+  nnz = sym_coo.nnz;
+  printf("\nSymmetrized COO Matrix (nnz=%ld):\n", nnz);
+
+  using value_idx = int;
+  using value_t   = float;
+  using size_type = size_t;
+
+  raft::sparse::op::coo_sort<float>(&sym_coo, stream);
+  auto row_ind = raft::make_device_vector<int>(handle, sym_coo.n_rows + 1);
+  raft::sparse::convert::sorted_coo_to_csr(&sym_coo, row_ind.data_handle(), stream);
+
+  const int one = sym_coo.nnz;
+  raft::copy(row_ind.data_handle() + row_ind.size() - 1, &one, 1, stream);
+
+  auto csr_structure = raft::make_device_compressed_structure_view<int, int, int>(
```

Comment:
You are ure using this here- why not also use the coo version above? That'll save us a lot of refactoring time in the future.